scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/simulation/ResumableSimulationDriver.java

```diff
@@ -165,7 +165,7 @@ public void simulateActivities(@NotNull Map<ActivityDirectiveId, ActivityDirecti
     resolved.get(null).sort(Comparator.comparing(Pair::getRight));
     final var earliestStartOffset = resolved.get(null).get(0);
 
-    if(earliestStartOffset.getRight().noLongerThan(curTime)){
+    if(earliestStartOffset.getRight().shorterThan(curTime)){
```

Comment:
TL;DR: Yes, simulation has always run everything up to but not including `curTime`

`curTime` is the time of the batch that the engine will resume at, which is always the first batch at a given time in the future of when the engine last stopped. In `simulateSchedule`, this is done by continuing the while loop if the tasks we're interested in are finished but there are still batches at the same time (`while (!allTaskFinished || delta.isZero())`). In `simulateUntil`, this is natural shakeout of the loop condition `!batch.offsetFromStart().longerThan(endTime)`, since all batches at a time will all obviously pass that check.

Reply:
Thanks - that helps point me in the right direction. Reading your explanation, and looking at the code, I'm not sure I reach the same conclusion. The code updates `curTime` only when the batch is about to be executed - if the `while` loop deems that the next batch is too far in the future, it leaves the value of `curTime` as the time of the last executed batch.

https://github.com/NASA-AMMOS/aerie/blob/c22b56cd74cf1a53917a071f0824978d6409103f/scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/simulation/ResumableSimulationDriver.java#L123-L125

https://github.com/NASA-AMMOS/aerie/blob/c22b56cd74cf1a53917a071f0824978d6409103f/scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/simulation/ResumableSimulationDriver.java#L247-L248

Reply:
Oh! Thanks for the catch. @adrienmaillard, how does the scheduler avoid any issues that may arise if a new task was placed between `curTime` and the start time of the batch the engine will resume on?

Reply:
That's a good question, I need to build a specific test to see what happens in that case.

Reply:
I think it can't happen because of how we maintain the current state of the simulation facade. We never have pending activities in the future because everytime we add an activity, we simulate until the end of all present activities (with or without restart). The `simulateUntil(t)` is only acting from the end of the last activity to `t`.

Reply:
What about with daemon tasks? Ie: Let's say that `curTime` is 10hrs, and that `batch` is a daemon task that's going to run at 12hrs. The next activity that is placed is at 11hrs. That would give us a task that's been placed between `curTime` and `batch.offsetFromStart`.

Reply:
> I think it can't happen

I'm not sure I quite follow the question we're answering here - I'll clarify what I'm concerned about. 

Let's say we previous simulated activity A, and have just been asked to simulate activity B which starts at the same time that activity A ended (note: B is not anchored to A, but merely happens to start at the end time of A).

Suppose that the final effect performed by activity A was to change a value from `Old Value` to `New Value`, and the first action taken by activity B is to read this value.

<img width="572" alt="Screenshot 2023-08-07 at 5 48 02 PM" src="https://github.com/NASA-AMMOS/aerie/assets/1189602/7b53ba4c-0285-4caf-835a-44f0558cb719">

What value does activity B see?

Aerie simulation semantics dictate that an activity in the plan that is scheduled to start at time `t` will not start causally after any effects that occur at time `t` - i.e. its first step will be concurrent with all other tasks that were scheduled to resume at time `t`. My concern is that the scheduler's simulation semantics will differ from Aerie's simulation semantics - possibly leading to subtle bugs

Reply:
This is the issue we were discussing with @Mythicaeda recently. Even though the semantics dictates something, there must be something in the implementation that make B execute after A because otherwise, a lot of activities placed by coexistence goals would not work as expected in our tests or Clipper.

Reply:
To address your point, I think that indeed, we are following different semantics in the scheduler. Maybe we should roll back on this change... We should probably open a ticket to modify how `CoexistenceGoal` works + documentation to tell people about the semantics of using `startsAt` (and the cost in simulation time). Maybe make some syntactic sugar with 'startsJustAfter' that would make the activity starts 1 microsecond after...

Reply:
I have created an issue #1087

Reply:
Reverted this change btw.

Reply:
I don't think reverting this change/the created ticket covers the other issue, which is that the scheduler can place activities in the gap between `curTime` and `batch.startOffset`, essentially placing work in the past of the engine (see [this comment](https://github.com/NASA-AMMOS/aerie/pull/1051#discussion_r1286488261) for an example case).

Reply:
@Mythicaeda this is probably worth another ticket.

Reply:
https://github.com/NASA-AMMOS/aerie/issues/1097

Reply:
Thank you!